hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapred/pipes/Application.java

```diff
@@ -266,4 +271,39 @@ public static String createDigest(byte[] password, String data)
     return SecureShuffleUtils.hashFromString(data, key);
   }
 
+  private class PingSocketCleaner extends Thread {
+    PingSocketCleaner(String name) {
+      super(name);
+    }
+
+    @Override
+    public void run() {
+      LOG.info("PingSocketCleaner started...");
+      while (true) {
+        Socket clientSocket = null;
+        try {
+          clientSocket = serverSocket.accept();
+          if (LOG.isDebugEnabled()) {
+            LOG.debug("Got one client socket...");
+          }
+          int readData = clientSocket.getInputStream().read();
```

Comment:
this is just for ping socket, since ping client won't send data but just make sure server can connect. we can do this because we know client behavior.  And yes, if other processes just connect server and send byte on purpose, the cleaner won't work well.

Reply:
and i think we can do like this:

```
         int readData = 0;
          while (readData != -1) {
            readData = clientSocket.getInputStream().read();
          }
          LOG.debug("close socket cause client has closed.");
          clientSocket.close();
```

Reply:
...and presumably a finally() round the read and use IOUtils.closeQuietly() for clientsocket so the case where it won't close because of some inner problem (such as it being closed:) don't trigger failures....